Override Configuration Using Environment Variables

michaelestrin opened the pull request:
Signed-off-by: Michael Estrin <m.estrin@dell.com>

#1450

tsconn23 commented on internal/core/command/init.go:
```diff
@@ -143,6 +143,8 @@ func initializeConfiguration(useRegistry bool, useProfile string) (*Configuratio
 		}
 	}
 
+	config.OverwriteFromEnvironment(configuration)
```
I admit I haven't looked at this really closely -- been taken up with other things. What I have looked at doesn't show a way to toggle this behavior. Maybe just point me in the right direction. We definitely don't want to do this by default.

michaelestrin replied:
The behavior can't be toggled per-se.  

If you set an environment variable named after a valid configuration field, the variable's value will overwrite the value already in configuration.  If you don't set appropriately named environment variables, the configuration will not be overridden.

Not sure if the concern is overhead to run the function.  It runs once during start up and time consumption is negligible.

tonyespy commented:
A couple of comments...

* During the last TSC F2F we said we weren't going to be working on any k8s related features... 

* There's a single commit for ~500 lines of diff, and the message just says "Override Configuration Using Environment Variables".  There's no commit message body which explains why this change was done. The body just contains the string "#1405". I think it's totally fine to such shorthand in github PR descriptions, however someone looking at this git repository on a local machine doesn't provide anyone any real context. I would suggest in commit messages that we always use full URLs for links to github issues.

 * While "."s in environment variables are technically legal, most shells don't support them

 * If I understand the proposal, environment vars take precedence over both local configuration and the registry. How will this impact the new work happening in the SMA for configuration?

akramtexas commented:
@tonyespy Good point, regarding your mentioning `"If I understand the proposal, environment vars take precedence over both local configuration and the registry. How will this impact the new work happening in the SMA for configuration?"` ➡️ @michaelestrin had reached out to me (in general, as in where the work per this PR "interfaces" with `System Management`). So yes, we need to have a conversation, going forward 👍

michaelestrin commented:
> There's no commit message body which explains why this change was done.

Absolutely correct.  Pushed reworded commit message.

lenny-intel commented:
I see you have refactored the config structs to remove maps for this to work via reflection for those settings. Why not have the override code handle maps via the data contained in them rather than ignoring them or refactoring them out?

michaelestrin commented:
> Why not have the override code handle maps via the data contained in them rather than ignoring them or refactoring them out?

Reflection doesn't support in-place modification of map elements (or their content).  Support for maps would have made the code significantly more complex.

@lenny-intel

lenny-intel commented:
@michaelestrin , I wasn't suggesting modification of the maps, just traversing them based on the contents. I think that complexity is worth not having the refactor out all the use of maps in the configs or ignoring them when left in the config.

michaelestrin commented:
@lenny-intel My original implementation did traverse them but I found that reflection doesn't support in-place modification of map elements _(**or their content**)_.   That is, I found it wasn't possible -- for instance -- to modify by reflection a string field in a sub-struct that was a field on a map element's struct.  While I could recreate the overall service-specific ConfigurationStruct and return it, I didn't feel the complexity involved in a generic solution was worth the effort.

FWIW, based on the Core WG feedback I received yesterday I've moved on to a different implementation -- one that does not use reflection -- and will be updating this PR with a detailed description shortly.

michaelestrin commented:
Based on the feedback received in the June 20th Core WG call, I modified my approach to solving this issue.  In addition to the feedback, I was influenced by the following:

1. The project's published Docker images use the registry by default.

1. The existing `Endpointer` contract and implementation (edgexfoundry/go-mod-core-contracts@v0.1.0/clients/endpointer.go) populates a service's address from configuration.  It then periodically polls the registry to retrieve that service's address and updates its local copy.  There is no clean way to extend the existing implementation to support environment variable overrides of service addresses as the service-instance level.

1. The existing `listenForConfigChanges()` implementation across the core services which listens for and updates the service's `ConfigurationStruct.Writable` struct when changes occur.  There is no clean way to extend the existing implementation to support environment variable overrides of `Writable` content changes made via the registry and pushed to a running service.  

1. Future system management agent work is planned to enable system management to push updated configuration to an individual service.  This is a variation of the problems described by the prior two points.  

1. Outside of a development environment, the registry service is likely to be a key system component. 

This PR has been updated to:

- Support environment variable overrides for config-seed to the `configuration.toml` content it uploads into registry.  An environment variable can be service-specific or generic.  

    A service-specific environment variable override would include the targeted service's name and would only affect configuration for that service; i.e. `core-metadata_Clients_logging_Host=<someAddr>`.  

    A generic environment variable override would not include a service name and would affect configuration across all services config-seed handles; i.e. `Registry_Host=<someAddr>`.

- Support environment variable override of a core service's RegistryInfo fields (i.e. host, port, and type) to allow a containerized service instance to be pointed at an alternate registry location.

@tsconn23 @tonyespy @lenny-intel @akramtexas

lenny-intel commented:
@michaelestrin , ahh not modify the map, but the field of the struct via the  map. Got it! :thumbsup:

akramtexas commented on internal/system/agent/init.go:
```diff
@@ -129,6 +129,7 @@ func initializeConfiguration(useRegistry bool, useProfile string) (*Configuratio
 	if err != nil {
 		return nil, err
 	}
+	configuration.Registry = config.OverwriteFromEnvironment(configuration.Registry)
```
Consider renaming the func `OverwriteFromEnvironment()` to `OverrideFromEnvironment()`.

michaelestrin replied:
Done.

akramtexas commented:
> ...
> 4. Future system management agent work is planned to enable system management to push updated configuration to an individual service.  This is a variation of the problems described by the prior two points.
> ... 
> @tsconn23 @tonyespy @lenny-intel @akramtexas

Referring to a point noted in the PR (i.e. `4. Future system management agent work is planned to enable system management to push updated configuration to an individual service. This is a variation of the problems described by the prior two points.`), please note that the SMA _will_, on occasion, though not always, be called upon to update configuration for _multiple_ service (e.g. update the setting with the key "Logging.EnableRemote" to take on the value "false").

akramtexas commented on internal/pkg/config/environment.go:
```diff
@@ -0,0 +1,42 @@
+)
+
+const (
+	envKeyPrefix = "edgex_registry_"
```
- As far as interacting with the Registry is concerned, please note that the SMA will (in all likelihood) use the go-mod-registry PutConfigurationValue() since it will not have the services config struct the is passed to `PutConfiguration`. And since the SMA will go directly to the Registry, rather than the service, the need exists for knowing the proper path in Consul for the (given) key.
- The Registry Client is created for a specific service with that services service key. That takes care of the base path.
- So we need to know the (trailing) `sub-path` for the specific configuration value.
- Here, in the context of the logic encoded in `environment.go`, would this logic have responsibility for handling that aspect? (i.e. "figure out the (trailing) `sub-path` for the specific configuration value").

michaelestrin replied:
The changes in this PR override the settings in a configuration.toml file.  They know nothing of the mapping from configuration.toml to registry.  The referenced source `internal/pkg/config/environment.go` simply host, port and type in a RegistryInfo struct that's contained within a service's ConfigurationStruct struct.

akramtexas replied:
That makes sense 👍 Thanks for confirming. I'm good with it.

michaelestrin commented:
> SMA _will_, on occasion, though not always, be called upon to update configuration for _multiple_ service

I see this as an orthogonal concern.  

config-seed overrides are there for system start up.  Once the system is underway, changes to the configuration stored in the registry (if it's being used) are outside the scope of the changes in this PR.

Do you concur @akramtexas ?

akramtexas commented:
> > SMA _will_, on occasion, though not always, be called upon to update configuration for _multiple_ service
> 
> I see this as an orthogonal concern.
> 
> config-seed overrides are there for system start up. Once the system is underway, changes to the configuration stored in the registry (if it's being used) are outside the scope of the changes in this PR.
> 
> Do you concur @akramtexas ?

@michaelestrin That was very helpfully-articulated. Yep, makes sense, and I concur 👍

anonymouse64 commented:
> * Support environment variable overrides for config-seed to the `configuration.toml` content it uploads into registry.  An environment variable can be service-specific or generic.

This kind of change is exactly what I was referring to in my second point here: https://github.com/edgexfoundry/edgex-go/pull/1348#pullrequestreview-238957299

This should enable us to drop the duplicated Redis configuration files from the docker image (see #1347 and #1359) and just override the necessary configuration.toml changes to use Redis with environment variables specified in the docker-compose.yml. 

I haven't reviewed this code at all (and don't intend to), I just wanted to chime in that I am +1 on this approach

akramtexas commented:
@michaelestrin I want to spend just a bit more time playing with the updated code in this PR, run the unit tests, etc., and will then approve the PR. Overall, so far, LGTM 👍

michaelestrin commented:
Updated to make @akramtexas suggested function name change.  Also corrected two instances where OverrideFromEnvironment() should have been called but wasn't.

tsconn23 commented on internal/pkg/config/environment_test.go:
```diff
@@ -0,0 +1,96 @@
+	registryInfo = OverrideFromEnvironment(registryInfo)
+
+	assert.Equal(t, registryInfo.Type, defaultTypeValue)
+}
```
@michaelestrin Question -- Have you looked at table driven tests at all in Go? It looks like these methods could utilize that technique, and you could thereby collapse these methods into one with a table containing a series of testable criteria and mocks. I think it's a little more concise. Example: https://github.com/edgexfoundry/edgex-go/blob/2cd11714e75ea178d29e77b27c6dfbe4b8618a15/internal/core/metadata/operators/device/add_test.go#L28

michaelestrin replied:
Recognize they're idiomatic.  I shy away from them because they tend to be harder for me to read/grok than the old school "one method, one test" approach.  Refactored per your implied request.

tsconn23 replied:
Yeah, I had that same cognitive clash. Look at them enough and you get used to it though.
